### PR TITLE
Melody parser class + tests

### PR DESCRIPTION
**Pull request by RedSparrowEU**
Initial DojoCode refactored.
New class MelodyParser + simple tests.
Works with any song file.

### CODE REVIEW

File: src/MelodyGenerator.js [hunk 1]
```diff
@@ -3,18 +3,122 @@ const MelodyQueue = require('./MelodyQueue.js');
...
+        switch (command[0]) {
+            case 'play':
+                this.parseCommandPlay(command);
+                break;
+            case 'define':
+                this.parseCommandDefine(command);
+                break;
+
+            default:
+                break;
```
Comment (firemark) [inline src/MelodyGenerator.js]: in future - maybe show warning is a better idea

File: src/MelodyGenerator.js [hunk 2]
```diff
@@ -3,18 +3,122 @@ const MelodyQueue = require('./MelodyQueue.js');
...
+        for (let i = 3; i < comm.length; i++) {
+            if (i % 2) {
+                if (this.defines[name] === undefined) this.defines[name] = [];
+                this.defines[name].push(comm[i]);
+            }
+        }
+    }
+
+    //add note - check if it is macro first
+    add(comm) {
```
Comment (firemark) [inline src/MelodyGenerator.js]: method name is ambiguous - maybe just `addNote`?

File: src/MelodyGenerator.js [hunk 3]
```diff
@@ -3,18 +3,122 @@ const MelodyQueue = require('./MelodyQueue.js');
...
+            }
+        }
+    }
+
+    //parse definitions (aka macros)
+    parseCommandDefine(comm) {
+        let name = comm[1];
+
+        for (let i = 3; i < comm.length; i++) {
+            if (i % 2) {
```
Comment (firemark) [inline src/MelodyGenerator.js]: what.

File: src/MelodyGenerator.js [hunk 4]
```diff
@@ -3,18 +3,122 @@ const MelodyQueue = require('./MelodyQueue.js');
...
-class MelodyGenerator
-{
+class MelodyGenerator {
     fromString(melodyString) {
         // put your code here
+        this.defines = [];
+        this.sounds = [];
+        let step1 = melodyString.replace(new RegExp(', ', "g"), ',').split("\n");
+
+        for (let i = 0; i < step1.length; i++) {
```
Comment (firemark) [inline src/MelodyGenerator.js]: I see many old for loops. Maybe foreach like `for(let step of step1)` or `step1.forEach(function)`

see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/forEach

Comment (RedSparrowEU) [reply src/MelodyGenerator.js]: For loop is maybe old, but the fastest loop in js. 
See: https://hackernoon.com/javascript-performance-test-for-vs-for-each-vs-map-reduce-filter-find-32c1113f19d7

File: src/MelodyGenerator.js (hunk 3 above)
Comment (RedSparrowEU) [reply src/MelodyGenerator.js]: What 'what' ? :)

File: src/MelodyGenerator.js (hunk 2 above)
Comment (RedSparrowEU) [reply src/MelodyGenerator.js]: Yes. addNote would be funtasctic.

File: src/MelodyGenerator.js (hunk 1 above)
Comment (RedSparrowEU) [reply src/MelodyGenerator.js]: Why not...

File: src/MelodyGenerator.js (hunk 3 above)
Comment (firemark) [reply src/MelodyGenerator.js]: What if code is wrong like `note(A,B) | | note(A,B)`?

File: src/MelodyGenerator.js (hunk 4 above)
Comment (firemark) [reply src/MelodyGenerator.js]: Don't optimize when you don't need.
http://wiki.c2.com/?RulesOfOptimization

File: src/MelodyGenerator.js (hunk 3 above)
Comment (RedSparrowEU) [reply src/MelodyGenerator.js]: Initial task was for pure code. Sure there is more to parse if not. :) Pipes more spaces in-between. I have solution for next task. Sure thing :)

File: src/MelodyGenerator.js (hunk 4 above)
Comment (RedSparrowEU) [reply src/MelodyGenerator.js]: You don't know what I need :) Here is my counterargument: https://www.phpied.com/extreme-javascript-optimization/ Especially at the bottom line...

Comment (firemark) [reply src/MelodyGenerator.js]: `You don't know what I need :) ` :fire:

File: src/MelodyGenerator.js (hunk 3 above)
Comment (tjaniczek) [reply src/MelodyGenerator.js]: Hi Greg, one small tip: you can traverse through odd numbers only via:
`for (let i = 3; i < comm.length; i += 2)`

Comment (RedSparrowEU) [reply src/MelodyGenerator.js]: Hi Tom, of course :)

**Pull request closed without merging**